File: sdks/go/tracing_test.go
package faro

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"
	"time"
)

// traceCaptureServer monta un test server que separa /v1/traces y /api/v1/ingest/logs.
// Diferente del captureServer existente en faro_test.go que solo maneja un endpoint.
type traceCaptureServer struct {
	mu     sync.Mutex
	traces []map[string]any
	logs   []map[string]any
	srv    *httptest.Server
}

func newTraceCaptureServer() *traceCaptureServer {
	c := &traceCaptureServer{}
	c.srv = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, _ := io.ReadAll(r.Body)
		var data map[string]any
		_ = json.Unmarshal(body, &data)
		c.mu.Lock()
		switch r.URL.Path {
		case "/v1/traces":
			c.traces = append(c.traces, data)
		case "/api/v1/ingest/logs":
			c.logs = append(c.logs, data)
		}
		c.mu.Unlock()
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{"ok":true}`))
	}))
	return c
}

func (c *traceCaptureServer) Close() { c.srv.Close() }

func (c *traceCaptureServer) waitForLocal(predicate func() bool, timeout time.Duration) bool {
	// El predicate suele llamar a allSpans()/etc que toman c.mu — NO lockear aquí
	// o creamos un deadlock (sync.Mutex no es reentrante en Go).
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		if predicate() {
			return true
		}
		time.Sleep(20 * time.Millisecond)
	}
	return false
}

func (c *traceCaptureServer) allSpans() []map[string]any {
	c.mu.Lock()
	defer c.mu.Unlock()
	var out []map[string]any
	for _, req := range c.traces {
		rs, _ := req["resourceSpans"].([]any)
		for _, rsv := range rs {
			rsmap, _ := rsv.(map[string]any)
			ss, _ := rsmap["scopeSpans"].([]any)
			for _, ssv := range ss {
				ssmap, _ := ssv.(map[string]any)
				spans, _ := ssmap["spans"].([]any)
				for _, sp := range spans {
					if m, ok := sp.(map[string]any); ok {
						out = append(out, m)
					}
				}
			}
		}
	}
	return out
}

func (c *traceCaptureServer) firstResourceAttrs() []any {
	c.mu.Lock()
	defer c.mu.Unlock()
	if len(c.traces) == 0 {
		return nil
	}
	rs, _ := c.traces[0]["resourceSpans"].([]any)
	if len(rs) == 0 {
		return nil
	}
	rsmap, _ := rs[0].(map[string]any)
	res, _ := rsmap["resource"].(map[string]any)
	attrs, _ := res["attributes"].([]any)
	return attrs
}

func getAttr(attrs []any, key string) string {
	for _, a := range attrs {
		m, ok := a.(map[string]any)
		if !ok {
			continue
		}
		if m["key"] != key {
			continue
		}
		v, _ := m["value"].(map[string]any)
		if s, ok := v["stringValue"].(string); ok {
			return s
		}
	}
	return ""
}

func newTestClient(t *testing.T, endpoint string) *Client {
	t.Helper()
	c, err := New(Options{
		Endpoint:      endpoint,
		Token:         "tk",
		Service:       "test",
		Environment:   "prod",
		Release:       "1.0.0",
		FlushInterval: 50 * time.Millisecond,
	})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	return c
}

func TestStartSpan_EmitsOTLPJSON(t *testing.T) {
	srv := newTraceCaptureServer()
	defer srv.Close()
	c := newTestClient(t, srv.srv.URL)

	_, span := c.StartSpan(context.Background(), "checkout", SpanOptions{
		Kind:       SpanKindServer,
		Attributes: map[string]any{"http.method": "POST"},
	})
	span.SetAttribute("user.id", 42)
	span.AddEvent("cache.miss", map[string]any{"key": "abc"})
	span.SetStatus(StatusOK, "")
	span.End()

	_ = c.Flush(2 * time.Second)
	if !srv.waitForLocal(func() bool { return len(srv.allSpans()) >= 1 }, 2*time.Second) {
		t.Fatal("span no llegó")
	}

	resAttrs := srv.firstResourceAttrs()
	if getAttr(resAttrs, "service.name") != "test" {
		t.Errorf("service.name = %q", getAttr(resAttrs, "service.name"))
	}
	if getAttr(resAttrs, "deployment.environment") != "prod" {
		t.Errorf("environment missing")
	}

	sp := srv.allSpans()[0]
	if sp["name"] != "checkout" {
		t.Errorf("name = %v", sp["name"])
	}
	if int(sp["kind"].(float64)) != int(SpanKindServer) {
		t.Errorf("kind = %v", sp["kind"])
	}
	tid := sp["traceId"].(string)
	if len(tid) != 32 {
		t.Errorf("traceId len = %d", len(tid))
	}
	sid := sp["spanId"].(string)
	if len(sid) != 16 {
		t.Errorf("spanId len = %d", len(sid))
	}
	if _, ok := sp["parentSpanId"]; ok {
		// el campo NO debe estar (omitempty); si está, no debe ser non-empty
		if s := sp["parentSpanId"].(string); s != "" {
			t.Errorf("root span con parentSpanId=%q", s)
		}
	}
	status, _ := sp["status"].(map[string]any)
	if int(status["code"].(float64)) != int(StatusOK) {
		t.Errorf("status.code = %v", status["code"])
	}
	spanAttrs, _ := sp["attributes"].([]any)
	if getAttr(spanAttrs, "user.id") != "42" {
		t.Errorf("user.id attr missing")
	}
	events, _ := sp["events"].([]any)
	if len(events) == 0 || events[0].(map[string]any)["name"] != "cache.miss" {
		t.Errorf("event cache.miss missing")
	}

	_ = c.Close(context.Background())
}

func TestStartSpan_ContextParent(t *testing.T) {
	srv := newTraceCaptureServer()
	defer srv.Close()
	c := newTestClient(t, srv.srv.URL)

	ctx, parent := c.StartSpan(context.Background(), "parent", SpanOptions{})
	_, child := c.StartSpan(ctx, "child", SpanOptions{})
	child.End()
	parent.End()
	_ = c.Flush(2 * time.Second)
	if !srv.waitForLocal(func() bool { return len(srv.allSpans()) >= 2 }, 2*time.Second) {
		t.Fatal("spans no llegaron")
	}

	var p, ch map[string]any
	for _, sp := range srv.allSpans() {
		switch sp["name"] {
		case "parent":
			p = sp
		case "child":
			ch = sp
		}
	}
	if ch["traceId"] != p["traceId"] {
		t.Errorf("child no hereda trace_id")
	}
	if ch["parentSpanId"] != p["spanId"] {
		t.Errorf("child.parentSpanId = %v, want %v", ch["parentSpanId"], p["spanId"])
	}

	_ = c.Close(context.Background())
}

func TestWithSpan_RecordsExceptionOnError(t *testing.T) {
	srv := newTraceCaptureServer()
	defer srv.Close()
	c := newTestClient(t, srv.srv.URL)

	err := c.WithSpan(context.Background(), "boom", func(ctx context.Context, span *Span) error {
		return fmt.Errorf("kaboom")
	}, SpanOptions{})
	if err == nil || !strings.Contains(err.Error(), "kaboom") {
		t.Fatalf("WithSpan no propagó el error: %v", err)
	}
	_ = c.Flush(2 * time.Second)
	if !srv.waitForLocal(func() bool { return len(srv.allSpans()) >= 1 }, 2*time.Second) {
		t.Fatal("span no llegó")
	}

	sp := srv.allSpans()[0]
	status := sp["status"].(map[string]any)
	if int(status["code"].(float64)) != int(StatusError) {
		t.Errorf("status no es ERROR: %v", status)
	}
	attrs := sp["attributes"].([]any)
	if getAttr(attrs, "exception.message") != "kaboom" {
		t.Errorf("exception.message missing")
	}

	_ = c.Close(context.Background())
}

func TestStartSpan_ForceRoot(t *testing.T) {
	srv := newTraceCaptureServer()
	defer srv.Close()
	c := newTestClient(t, srv.srv.URL)

	ctx, parent := c.StartSpan(context.Background(), "outer", SpanOptions{})
	_, child := c.StartSpan(ctx, "forced-root", SpanOptions{Parent: SpanParent{ForceRoot: true}})
	child.End()
	parent.End()
	_ = c.Flush(2 * time.Second)
	if !srv.waitForLocal(func() bool { return len(srv.allSpans()) >= 2 }, 2*time.Second) {
		t.Fatal("spans no llegaron")
	}
	var p, ch map[string]any
	for _, sp := range srv.allSpans() {
		switch sp["name"] {
		case "outer":
			p = sp
		case "forced-root":
			ch = sp
		}
	}
	if ch["traceId"] == p["traceId"] {
		t.Errorf("ForceRoot debería romper la herencia de trace_id")
	}

	_ = c.Close(context.Background())
}

func TestStartSpan_TraceparentParent(t *testing.T) {
	srv := newTraceCaptureServer()
	defer srv.Close()
	c := newTestClient(t, srv.srv.URL)

	tp := "00-" + strings.Repeat("a", 32) + "-" + strings.Repeat("b", 16) + "-01"
	_, span := c.StartSpan(context.Background(), "from-tp", SpanOptions{
		Parent: SpanParent{Traceparent: tp},
	})
	span.End()
	_ = c.Flush(2 * time.Second)
	if !srv.waitForLocal(func() bool { return len(srv.allSpans()) >= 1 }, 2*time.Second) {
		t.Fatal("span no llegó")
	}

	sp := srv.allSpans()[0]
	if sp["traceId"] != strings.Repeat("a", 32) {
		t.Errorf("traceId no del traceparent: %v", sp["traceId"])
	}
	if sp["parentSpanId"] != strings.Repeat("b", 16) {
		t.Errorf("parentSpanId no del traceparent: %v", sp["parentSpanId"])
	}

	_ = c.Close(context.Background())
}

func TestLogContext_AutoCorrelates(t *testing.T) {
	srv := newTraceCaptureServer()
	defer srv.Close()
	c := newTestClient(t, srv.srv.URL)

	ctx, span := c.StartSpan(context.Background(), "handler", SpanOptions{})
	c.InfoContext(ctx, "procesando", map[string]any{"foo": "bar"})
	span.End()
	_ = c.Flush(2 * time.Second)
	if !srv.waitForLocal(func() bool {
		srv.mu.Lock()
		defer srv.mu.Unlock()
		return len(srv.logs) >= 1
	}, 2*time.Second) {
		t.Fatal("log no llegó")
	}

	srv.mu.Lock()
	logBatch := srv.logs[0]
	srv.mu.Unlock()
	logs := logBatch["logs"].([]any)
	log0 := logs[0].(map[string]any)
	if log0["trace_id"] != span.TraceID() {
		t.Errorf("log.trace_id = %v, want %v", log0["trace_id"], span.TraceID())
	}
	if log0["span_id"] != span.SpanID() {
		t.Errorf("log.span_id = %v, want %v", log0["span_id"], span.SpanID())
	}

	_ = c.Close(context.Background())
}

func TestSpan_TraceparentFormat(t *testing.T) {
	c, _ := New(Options{
		Endpoint: "http://127.0.0.1:1", Token: "t", Service: "s",
		FlushInterval: time.Hour,
	})
	defer c.Close(context.Background())
	_, span := c.StartSpan(context.Background(), "x", SpanOptions{})
	tp := span.Traceparent()
	if !strings.HasPrefix(tp, "00-") || !strings.HasSuffix(tp, "-01") {
		t.Errorf("traceparent format inválido: %s", tp)
	}
	parts := strings.Split(tp, "-")
	if len(parts) != 4 || len(parts[1]) != 32 || len(parts[2]) != 16 {
		t.Errorf("traceparent parts inválidos: %v", parts)
	}
}

func TestSpan_RecordException(t *testing.T) {
	srv := newTraceCaptureServer()
	defer srv.Close()
	c := newTestClient(t, srv.srv.URL)

	_, span := c.StartSpan(context.Background(), "op", SpanOptions{})
	span.RecordException(fmt.Errorf("bad input"))
	span.End()
	_ = c.Flush(2 * time.Second)
	if !srv.waitForLocal(func() bool { return len(srv.allSpans()) >= 1 }, 2*time.Second) {
		t.Fatal("span no llegó")
	}
	sp := srv.allSpans()[0]
	status := sp["status"].(map[string]any)
	if int(status["code"].(float64)) != int(StatusError) {
		t.Errorf("status no es ERROR")
	}
	attrs := sp["attributes"].([]any)
	if getAttr(attrs, "exception.message") != "bad input" {
		t.Errorf("exception.message missing")
	}

	_ = c.Close(context.Background())
}

func TestSpan_End_Idempotent(t *testing.T) {
	srv := newTraceCaptureServer()
	defer srv.Close()
	c := newTestClient(t, srv.srv.URL)
	_, span := c.StartSpan(context.Background(), "once", SpanOptions{})
	span.End()
	span.End() // segunda llamada NO debe encolar otro
	span.End()
	_ = c.Flush(2 * time.Second)
	if !srv.waitForLocal(func() bool { return len(srv.allSpans()) >= 1 }, 1*time.Second) {
		t.Fatal("primer End debería haber encolado")
	}
	time.Sleep(200 * time.Millisecond)
	if n := len(srv.allSpans()); n != 1 {
		t.Errorf("End no es idempotente: %d spans", n)
	}

	_ = c.Close(context.Background())
}

File: sdks/node/test/tracing.test.mjs
/**
 * Tests del SDK de tracing — invariantes públicas (v0.2.0):
 *   1. startSpan / end → flush manda OTLP/JSON a /v1/traces con service.name correcto
 *   2. parent context explícito: hijo hereda trace_id y setea parentSpanId
 *   3. withSpan: cierra solo y propaga ERROR si fn lanza
 *   4. Context manager: spans anidados sin pasar context (auto-parent)
 *   5. logs dentro de withSpan auto-heredan trace_id/span_id
 *   6. traceparent() formato W3C
 *   7. recordException popula attributes y status=ERROR
 *
 * NOTA: en v0.2.0 el tracing está respaldado por @opentelemetry/sdk-trace-node.
 * El BatchSpanProcessor se drena via faro.flush() (que llama a forceFlush()).
 * No miramos la cola interna — el proceso de export es opaque.
 */

import { test } from 'node:test';
import assert from 'node:assert/strict';
import http from 'node:http';

const faro = await import('../dist/index.js');

function startServer(handler) {
  return new Promise((resolve) => {
    const server = http.createServer((req, res) => {
      let body = '';
      req.on('data', (c) => (body += c));
      req.on('end', () => handler(req, res, body));
    });
    server.listen(0, () => resolve({ server, port: server.address().port }));
  });
}

/** Acumula los bodies recibidos en /v1/traces para inspección. */
function makeTraceCollector() {
  const captured = [];
  let lastPath = null;
  const handler = (req, res, body) => {
    lastPath = req.url;
    if (req.url === '/v1/traces' && body) {
      try {
        captured.push(JSON.parse(body));
      } catch {
        // ignore
      }
    }
    res.writeHead(200, { 'content-type': 'application/json' });
    res.end('{"partialSuccess":{}}');
  };
  return {
    handler,
    captured,
    lastPath: () => lastPath,
    /** Devuelve todos los spans recibidos planos. */
    allSpans: () => captured.flatMap((req) =>
      (req.resourceSpans ?? []).flatMap((rs) =>
        (rs.scopeSpans ?? []).flatMap((ss) => ss.spans ?? []),
      ),
    ),
    /** Devuelve los resource attributes del primer batch. */
    firstResourceAttrs: () => captured[0]?.resourceSpans?.[0]?.resource?.attributes ?? [],
  };
}

function getStringAttr(attrs, key) {
  const a = attrs?.find((x) => x.key === key);
  return a?.value?.stringValue ?? a?.value?.intValue ?? a?.value?.boolValue;
}

test('startSpan: emite OTLP/JSON a /v1/traces con service.name correcto', async () => {
  const c = makeTraceCollector();
  const { server, port } = await startServer(c.handler);

  const f = faro.init({
    endpoint: `http://127.0.0.1:${port}`,
    token: 'tk', service: 'trace-test',
    installGlobalHandlers: false, flushIntervalMs: 100_000,
    environment: 'prod', release: '1.2.3',
    diag: () => {},
  });
  try {
    const span = f.startSpan('checkout', {
      kind: 'SERVER',
      attributes: { 'http.method': 'POST' },
    });
    span.setAttribute('user.id', 42);
    span.addEvent('cache.miss', { attributes: { key: 'abc' } });
    span.setStatus('OK');
    span.end();
    await f.flush();

    assert.equal(c.lastPath(), '/v1/traces', 'pega a /v1/traces');
    const resAttrs = c.firstResourceAttrs();
    assert.equal(getStringAttr(resAttrs, 'service.name'), 'trace-test');
    assert.equal(getStringAttr(resAttrs, 'service.version'), '1.2.3');
    assert.ok(
      getStringAttr(resAttrs, 'deployment.environment') === 'prod' ||
      getStringAttr(resAttrs, 'deployment.environment.name') === 'prod',
      'environment presente en el resource',
    );

    const spans = c.allSpans();
    const sp = spans.find((s) => s.name === 'checkout');
    assert.ok(sp, 'span "checkout" presente');
    assert.equal(sp.kind, 2, 'SERVER=2');
    assert.match(sp.traceId, /^[0-9a-f]{32}$/);
    assert.match(sp.spanId, /^[0-9a-f]{16}$/);
    assert.ok(!sp.parentSpanId || sp.parentSpanId === '', 'root span no debe tener parentSpanId');
    assert.equal(sp.status?.code, 1, 'OK=1');
    assert.equal(getStringAttr(sp.attributes, 'user.id'), '42');
    assert.equal(sp.events?.[0]?.name, 'cache.miss');
  } finally {
    await f.close(2000);
    server.close();
  }
});

test('parent context explícito: hijo hereda trace_id y setea parentSpanId', async () => {
  const c = makeTraceCollector();
  const { server, port } = await startServer(c.handler);

  const f = faro.init({
    endpoint: `http://127.0.0.1:${port}`, token: 'tk', service: 't-parent',
    installGlobalHandlers: false, flushIntervalMs: 100_000, diag: () => {},
  });
  try {
    const parent = f.startSpan('parent');
    const parentCtx = parent.spanContext();
    const child = f.startSpan('child', { parent: parentCtx });
    child.end();
    parent.end();
    await f.flush();

    const spans = c.allSpans();
    const parentSp = spans.find((s) => s.name === 'parent');
    const childSp = spans.find((s) => s.name === 'child');
    assert.ok(parentSp && childSp);
    assert.equal(childSp.traceId, parentSp.traceId, 'hijo hereda trace_id');
    assert.equal(childSp.parentSpanId, parentSp.spanId);
  } finally {
    await f.close(2000);
    server.close();
  }
});

test('withSpan: cierra solo y propaga ERROR si fn lanza', async () => {
  const c = makeTraceCollector();
  const { server, port } = await startServer(c.handler);

  const f = faro.init({
    endpoint: `http://127.0.0.1:${port}`, token: 'tk', service: 't-throws',
    installGlobalHandlers: false, flushIntervalMs: 100_000, diag: () => {},
  });
  try {
    await assert.rejects(
      f.withSpan('boom', async () => { throw new Error('kaboom'); }),
      /kaboom/,
    );
    await f.flush();
    const sp = c.allSpans().find((s) => s.name === 'boom');
    assert.ok(sp);
    assert.equal(sp.status?.code, 2, 'ERROR=2');
    assert.ok(getStringAttr(sp.attributes, 'exception.type'));
    assert.ok(getStringAttr(sp.attributes, 'exception.message'));
  } finally {
    await f.close(2000);
    server.close();
  }
});

test('Context manager: spans anidados sin pasar context (auto-parent)', async () => {
  const c = makeTraceCollector();
  const { server, port } = await startServer(c.handler);

  const f = faro.init({
    endpoint: `http://127.0.0.1:${port}`, token: 'tk', service: 't-nested',
    installGlobalHandlers: false, flushIntervalMs: 100_000, diag: () => {},
  });
  try {
    await f.withSpan('outer', async () => {
      await f.withSpan('inner', async () => {
        const active = f.activeSpan();
        assert.ok(active, 'activeSpan no debe ser null dentro de withSpan');
      });
    });
    await f.flush();
    const spans = c.allSpans();
    const outer = spans.find((s) => s.name === 'outer');
    const inner = spans.find((s) => s.name === 'inner');
    assert.ok(outer && inner);
    assert.equal(inner.traceId, outer.traceId);
    assert.equal(inner.parentSpanId, outer.spanId);
  } finally {
    await f.close(2000);
    server.close();
  }
});

test('logs dentro de withSpan auto-heredan trace_id/span_id', async () => {
  const c = makeTraceCollector();
  let logsBody = null;
  const { server, port } = await startServer((req, res, body) => {
    if (req.url === '/api/v1/ingest/logs') {
      logsBody = JSON.parse(body);
      res.writeHead(200); res.end('{}');
      return;
    }
    c.handler(req, res, body);
  });

  const f = faro.init({
    endpoint: `http://127.0.0.1:${port}`, token: 'tk', service: 't-corr',
    installGlobalHandlers: false, flushIntervalMs: 100_000, diag: () => {},
  });
  try {
    let spanCtx;
    await f.withSpan('handler', async (span) => {
      spanCtx = span.spanContext();
      f.info('procesando request', { foo: 'bar' });
    });
    await f.flush();
    const log = logsBody?.logs?.[0];
    assert.ok(log, 'log presente');
    assert.equal(log.trace_id, spanCtx.trace_id, 'trace_id hereda del span activo');
    assert.equal(log.span_id, spanCtx.span_id, 'span_id hereda del span activo');
  } finally {
    await f.close(2000);
    server.close();
  }
});

test('traceparent(): formato W3C válido', async () => {
  // Sin red, sólo verificamos el formato del header.
  const f = faro.init({
    endpoint: 'http://127.0.0.1:1', token: 'tk', service: 't-tp',
    installGlobalHandlers: false, flushIntervalMs: 100_000,
    enableTracing: true, diag: () => {},
  });
  try {
    const span = f.startSpan('x');
    const tp = span.traceparent();
    assert.match(tp, /^00-[0-9a-f]{32}-[0-9a-f]{16}-(00|01)$/);
    span.end();
  } finally {
    await f.close(50);
  }
});

test('recordException: setea ERROR + exception.* attrs sin re-lanzar', async () => {
  const c = makeTraceCollector();
  const { server, port } = await startServer(c.handler);

  const f = faro.init({
    endpoint: `http://127.0.0.1:${port}`, token: 'tk', service: 't-exc',
    installGlobalHandlers: false, flushIntervalMs: 100_000, diag: () => {},
  });
  try {
    const span = f.startSpan('op');
    span.recordException(new TypeError('bad input'));
    span.end();
    await f.flush();
    const sp = c.allSpans().find((s) => s.name === 'op');
    assert.ok(sp);
    assert.equal(sp.status?.code, 2);
    assert.equal(getStringAttr(sp.attributes, 'exception.type'), 'TypeError');
    assert.equal(getStringAttr(sp.attributes, 'exception.message'), 'bad input');
  } finally {
    await f.close(2000);
    server.close();
  }
});

test('enableTracing:false omite la inicialización OTel', async () => {
  // Spans creados quedan no-op (traceId=000...000). El test verifica que init no rompe.
  const f = faro.init({
    endpoint: 'http://127.0.0.1:1', token: 'tk', service: 't-disabled',
    installGlobalHandlers: false, flushIntervalMs: 100_000,
    enableTracing: false, diag: () => {},
  });
  try {
    const span = f.startSpan('x');
    const sc = span.spanContext();
    // Tracer no-op del global → traceId todo ceros. Lo importante: no rompe.
    assert.match(sc.trace_id, /^[0-9a-f]{32}$/);
    span.end();
  } finally {
    await f.close(50);
  }
});
